hash-tables/overview-of-hash-tables.md

```diff
@@ -0,0 +1,153 @@
+
+### !end-callout
+
+## Collisions
+
+There is an enormous number of possible keys and a limited-size array. At some point, a hash function will result in some keys being mapped to the same index. When this occurs, it is called a _collision_.
+
+### Example Collision
+
+Let's use the same hash function from before:
+
+```python
+def hash_code(key):
+  return len(str(key)) % len(hash_array)
+```
+
+Imagine a hash table that stores "name"-"job title" key-value pairs. The hash table's internal `hash_array` has a length of `15`. Imagine inserting the following pairs, which have the following indices after applying the above hash function:
+
+| Key (Name) | Value (Job Title)     | Index (`hash_code(key)`) |
+| ---------- | --------------------- | ------------------------ |
+| `"dan"`    | `"Lead Instructor"`   | `3`                      |
+| `"chris"`  | `"CS Fun Instructor"` | `5`                      |
+| `"dee"`    | `"Lead Instructor"`   | `3`                      |
+
+If we insert Dan first, and then eventually Dee, our hash function will attempt to add Dee at the index `3`, even though Dan already exists there.
+
+### Resolving Collisions
+
+There are many ways to resolve a collision! One way is to make each element of the hash array a collection itself, and then store new elements into the collection.
+
+### Example Resolution
+
+If the values are stored in an array themselves, we can end up with a hash which looks like this:
+
+![Bad hash function example](../assets/hash-tables_hash-tables_collision.png)
+
+In this example, we resolve collisions by creating an array at each index of the main array, then adding the key-value pair to the end of the list.
+
+### !callout-info
+
+## Effective Hash Functions
+
+What would happen if we inserted 600 items, and our hash function happened to insert all 600 items into the same key? If the hash function does not adequately spread values out over the length of the array, we'll have long chains of key-value pairs and an O(n) worst case runtime to find items.
+
+Effective hash functions attempt to spread all possible values over the entire data structure to avoid assigning multiple keys to the same index. Most libraries use well-proven hashing functions which normally prevent high numbers of collisions.
+
+### !end-callout
+
+## Summary
+
+Hash tables are a data structure which allow quick lookup of a value by using a key. Each key is mapped to an index of an internal array with a specialized method called a _hash function_. The average-case lookup is O(1) and the worst-case is O(n). However, most good implementations use a _hash function_ which attempt to uniformly distribute elements and prevent collisions.
```

Comment:
```suggestion
Hash tables are a data structure which allow quick lookup of a value by using a key. Each key is mapped to an index of an internal array with a specialized method called a _hash function_. The average-case lookup is O(1) and the worst-case is O(n). However, most good implementations use a _hash function_ which attempts to uniformly distribute elements and prevent collisions.
```

Reply:
Just feel that we should explicitly close with the point that we can effectively treat hash tables as O(1) for lookup etc despite their worst-case.

```suggestion
Hash tables are a data structure which allow quick lookup of a value by using a key. Each key is mapped to an index of an internal array with a specialized method called a _hash function_. The average-case lookup is O(1) and the worst-case is O(n). However, most good implementations use a _hash function_ which attempts to uniformly distribute elements and prevent collisions.

This means that for _practical purposes_ we can treat the time complexity of lookup, deletion and insertion as O(1), despite the worst-case O(n) time complexity.  This is because, the implementations we use in languages like Python, so rarely encounter worst-case scenarios.
```